core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala

```diff
@@ -300,8 +300,8 @@ private[spark] class ExternalSorter[K, V, C](
         val partitionId = elem._1._1
         val key = elem._1._2
         val value = elem._2
-        writer.write(key)
-        writer.write(value)
+        writer.writeKey(key)
+        writer.writeValue(value)
```

Comment:
@rxin An alternative way to make the specialized SQL serializer work in most of the cases is to only change 
```
writer.write(key)
writer.write(value)
```
to 
```
writer.write((key, value))
```